### ISSUE TITLE
RPC: limit of number of connections?

### ISSUE THREAD

**Issue by pw374** [tezedge/tezedge#914]
Hi, is there a limit in the number of RPC requests that tezedge is willing to have open?

It seems (I'm not 100% sure (yet)) that tezedge will reject RPC connections when there are a number already running.

In my particular use case(*), responding to RPCs (without failing) is more important than anything else (e.g., P2P stuff).

(*) using tezedge to feed tezos-indexer

**Comment by tizoc** [comment]
@pw374 hello, can you give me some examples of the RPCs that are failing?

**Comment by pw374** [comment]
basically everything, it seems:
```
phil@amd:~/indexer$ curl -v http://localhost:8720/version
*   Trying 127.0.0.1:8720...
* Connected to localhost (127.0.0.1) port 8720 (#0)
> GET /version HTTP/1.1
> Host: localhost:8720
> User-Agent: curl/7.74.0
> Accept: */*
> 
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

```
phil@amd:~/indexer$ curl -v http://localhost:8720/chains/main/blocks/0/hash
*   Trying 127.0.0.1:8720...
* Connected to localhost (127.0.0.1) port 8720 (#0)
> GET /chains/main/blocks/0/hash HTTP/1.1
> Host: localhost:8720
> User-Agent: curl/7.74.0
> Accept: */*
> 
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

I have multiple indexers that are still running using that tezedge node and they don't seem to have crashed. Or maybe they have crashed silently and are actually stuck... I'm gonna have a closer look.

**Comment by tizoc** [comment]
Interesting, `/version` in particular is very lightweight, so even under load it is an endpoint that should be able to quickly produce a response.

Do you see anything weird in the tezedge logs? can you upload them for me to take a look?

**Comment by pw374** [comment]
Oh...
```
Jan 05 18:50:32.259 ERRO Failed to execute RPC function, reason: UnexpectedError { reason: "IPC error: Send error: Transport endpoint is not connected (os error 107)" }
```

I guess I gotta restart it...

**Comment by pw374** [comment]
I restarted it, got /version to work... a few times.

Now it's back to `Failed to execute RPC function `... :/

**Comment by tizoc** [comment]
@pw374 I don't think so. But the full log would be very useful, something needs to have happened before for that failure to occur. And that error shouldn't happen for `/version` because it doesn't need to contact the protocol runner process in any way.

How much RAM do you have? which context storage backend are you using? and what do you see in htop? how is memory usage?

**Comment by tizoc** [comment]
Also, does your node happen to be at block 1916927? Because the next block takes a while to process (it is the switch from Granada go Hangzhou, and it is a very expensive migration that can take roughly 30 mins to complete)

**Comment by pw374** [comment]
Level:
```
phil@amd:~/indexer$ curl -s http://localhost:8721/chains/main/blocks/head/header|jq .level
2006093
```

Logs:
https://gitlab.com/nomadic-labs/tezos-indexer/-/snippets/2230751


RAM:
```
phil@amd:~/indexer$ free -h
              total        used        free      shared  buff/cache   available
Mem:          125Gi        41Gi        16Gi        16Gi        67Gi        66Gi
Swap:            0B          0B          0B
```

**Comment by tizoc** [comment]
@pw374 ok, it is very weird, I don't see any error before such failure, but I would expect something.

Please shut down the node, and make sure that there aren't any `protocol-runner` processes still alive, if there are any, kill them and then try starting the node again.

Btw, what is the output of `ulimit -a` ?

**Comment by pw374** [comment]
```
$ ulimit -a
real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) 0
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 0
file size                   (blocks, -f) unlimited
pending signals                     (-i) 514752
max locked memory           (kbytes, -l) 16480893
max memory size             (kbytes, -m) unlimited
open files                          (-n) 1024
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 0
stack size                  (kbytes, -s) 8192
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) 514752
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited
```

I don't think there can be any `protocol-runner` still alive after the docker container is killed. Can there?

**Comment by pw374** [comment]
Tried again, failed again (a tiny bit less fast though)
https://gitlab.com/nomadic-labs/tezos-indexer/-/snippets/2230765

**Comment by pw374** [comment]
Do you need `ulimit -a` 's output from inside the docker container?!

If so, I'm not sure how to get that. (That docker image is particularly lightweight, it seems...)

**Comment by tizoc** [comment]
@pw374 ok, try adding this to your docker command to increase the ulimit value for files to see if that helps:

`--ulimit nofile=20000:20000`

**Comment by pw374** [comment]
> @pw374 ok, try adding this to your docker command to increase the ulimit value for files to see if that helps:
> 
> `--ulimit nofile=20000:20000`

Not 100% sure, but I might have made it even worse... 🥲 (if not, it's not much better anyways)

```
Jan 05 22:07:08.997 ERRO Failed to execute RPC function, reason: UnexpectedError { reason: "IPC error: Send error: Transport endpoint is not connected (os error 107)" }
```
such errors happened pretty much immediately.

Tried `--ulimit nofile=200000:200000` too. Same...

**Comment by pw374** [comment]
When I launch my indexers, there are `48 n` indexers querying the node in parallel.
(n being somewhere between 0 and 2 or 3)

It's not that huge, is it? octez hasn't had any issue dealing with the number of queries, however since it doesn't do much parallelism, one octez node will not like having that many queries. Therefore I usually split the queries to several nodes (the most often, it was 7, recently I've experimented with up to 40+ nodes).

Note that when I do `2 n` instead of `48 n`, tezedge doesn't complain.

**Comment by tizoc** [comment]
We will probably need to add some internal rate limiting. We run a separate process for protocol related operations, and when an RPC needs to run code from the protocol, the shell connects to that process and ask it to run protocol-related code. When this happens a new connection is created, which in this case may be too much.

**Comment by pw374** [comment]
I'm going to abandon TezEdge for now, since even without issues, it does not outperform multi-octez (i.e., multiple instances of octez, in order to have some multicore ability) on my machine for my particular use case.

Please let me know if you'd like me to try a patch on this resource limitation issue, I'll be happy to run it and give you feedbacks.

Also, please let me know if there's any update on performance.

**Comment by tizoc** [comment]
@pw374 the current version of TezEdge doesn't execute protocol code in parallel (it used to, and it will do so again in the future), so for protocol-heavy code (and if it has not been reimplemented in the Rust shell, like `baking_rights` for example) it will probably not be able to outperform multiple instances of Octez until such parallelism is re-implemented. 

I would be interested in seeing the kind of requests you are making. Do you have any test program we can use to reproduce the workload?
